### PR TITLE
[bpo-28414] Make all hostnames in SSL module IDN A-labels

### PR DESCRIPTION
**Pull request by tiran**
Alternative approach to PR #3010

- Really all server_hostname attributes are IDN A-labels
- SSLContext.set_servername_callback argument is IDN A-label, too
- Replace all externally replaced certs with proper certs from internal test helper

https://bugs.python.org/issue28414

### CODE REVIEW

File: Doc/whatsnew/3.7.rst [hunk 1]
```diff
@@ -441,6 +441,19 @@ can be set within the scope of a group.
...
+---
+
+Added support for validating server certificates containing
+internationalized domain names (IDNs). As part of this change, the
+:attr:`ssl.SSLSocket.server_hostname` and
+:attr:`ssl.SSLObject.server_hostname` attributes now store the
+expected hostname in A-label form (``"xn--pythn-mua.org"``), rather
+than the U-label form (``"pythön.org"``). The server name argument to
+:meth:`ssl.SSLContext.set_servername_callback`` is an A-label instead of
+an U-labe, too. (Contributed by Nathaniel J. Smith and Christian Heimes
```
Comment (alex) [inline Doc/whatsnew/3.7.rst]: "labe"

Comment (alex) [reply Doc/whatsnew/3.7.rst]: (also not sure what the deal with github rendering all this with underlines is; potential reST issue?)

File: Lib/test/test_asyncio/test_events.py [hunk 2]
```diff
@@ -1181,6 +1181,8 @@ def test_create_unix_server_ssl_verified(self):
         server.close()
         self.loop.run_until_complete(proto.done)
 
+    maxDiff = None
```
Comment (alex) [inline Lib/test/test_asyncio/test_events.py]: Unrelated?

Comment (tiran) [reply Lib/test/test_asyncio/test_events.py]: Remnants of debugging failing tests

File: Doc/whatsnew/3.7.rst (hunk 1 above)
Comment (tiran) [reply Doc/whatsnew/3.7.rst]: It's caused by a supernumerous backtick. I have fixed both issues.

**Review by asvetlov — APPROVED** [review]

**Comment by njsmith** [comment]
I did a quick skim and the code seems fine to me, but... I do think that for `set_servername_callback`, instead of just breaking things we maybe we need to rename it and add a compatibility shim. Quoting from https://mail.python.org/pipermail/python-dev/2017-December/151527.html:

> On the server, the obvious fix would be to start passing
> A-label-encoded names to the servername_callback, instead of
> U-label-encoded names. Unfortunately, this is a bit trickier, because
> this *has* historically worked (AFAIK) for IDNA names, so long as they
> didn't use one of the four magic characters who changed meaning
> between IDNA 2003 and IDNA 2008. But we do still need to do something.
> For example, right now, it's impossible to use the ssl module to
> implement a web server at https://straße.de, because incoming
> connections will use SNI to say that they expect a cert for
> "xn--strae-oqa.de", and then the ssl module will freak out and throw
> an exception instead of invoking the servername callback.
>
> It's ugly, but probably the simplest thing is to add a new function
> like `set_servername_callback2` that uses the A-label, and then redefine
> `set_servername_callback` as a deprecated compatibility shim:
> 
> ```python
> def set_servername_callback(self, cb):
>     def shim_cb(sslobj, servername, sslctx):
>         if servername is not None:
>             servername = servername.encode("ascii").decode("idna")
>         return cb(sslobj, servername, sslctx)
>     self.set_servername_callback2(shim_cb)
> ```
>
> We can bikeshed what the new name should be. Maybe `set_sni_callback`?
> or `set_server_hostname_callback`, since the corresponding client-mode
> argument is `server_hostname`?

**Comment by tiran** [comment]
Meh! :(

Let's make it PEP 543 compatible and add ``sni_callback`` property. The Python subclass can provide your shim cb. Thanks for proposing it!

File: Lib/ssl.py [hunk 3]
```diff
@@ -389,6 +400,20 @@ def set_npn_protocols(self, npn_protocols):
 
         self._set_npn_protocols(protos)
 
+    def set_servername_callback(self, server_name_callback):
+        if server_name_callback is None:
+            self.sni_callback = None
+        else:
+            if not hasattr(server_name_callback, '__call__'):
+                raise TypeError("not a callable object")
```
Comment (njsmith) [inline Lib/ssl.py]: Why not `if not callable(server_name_callback)`?

File: Lib/ssl.py [hunk 4]
```diff
@@ -389,6 +400,20 @@ def set_npn_protocols(self, npn_protocols):
...
+
+            def shim_cb(sslobj, servername, sslctx):
+                if servername is not None:
+                    servername = servername.encode("ascii").decode("idna")
```
Comment (njsmith) [inline Lib/ssl.py]: It would be simpler to use `_encode_hostname` here, no?

File: Lib/ssl.py [hunk 5]
```diff
@@ -355,13 +355,20 @@ def __new__(cls, protocol=PROTOCOL_TLS, *args, **kwargs):
         self = _SSLContext.__new__(cls, protocol)
         return self
 
-    def __init__(self, protocol=PROTOCOL_TLS):
-        self.protocol = protocol
+    def _encode_hostname(self, hostname):
```
Comment (njsmith) [inline Lib/ssl.py:358]: Maybe make this a top-level utility function, instead of a method?

File: Modules/_ssl.c [hunk 6]
```diff
@@ -4217,8 +4248,12 @@ static PyGetSetDef context_getsetlist[] = {
                        (setter) set_check_hostname, NULL},
     {"_host_flags", (getter) get_host_flags,
                     (setter) set_host_flags, NULL},
+    {"sni_callback", (getter) get_sni_callback,
+                       (setter) set_sni_callback, PySSLContext_sni_callback_doc},
     {"options", (getter) get_options,
                 (setter) set_options, NULL},
+    {"protocol", (getter) get_protocol,
+                 NULL, NULL},
```
Comment (njsmith) [inline Modules/_ssl.c:4245]: Why is this new `protocol` attribute in this PR? Should it be documented or anything?

File: Modules/_ssl.c [hunk 7]
```diff
@@ -4005,45 +4009,62 @@ _servername_callback(SSL *s, int *al, void *args)
...
+        if (result == Py_None) {
+            ret = SSL_TLSEXT_ERR_OK;
+        }
+        else if (PySSLContext_Check(result)) {
+            if (ssl->ctx != (PySSLContext *)result) {
+                /* if result is not sock.context: socket.context = result */
+                Py_INCREF(result);
+                Py_SETREF(ssl->ctx, (PySSLContext *)result);
+                SSL_set_SSL_CTX(ssl->ssl, ssl->ctx->ctx);
+            }
```
Comment (njsmith) [inline Modules/_ssl.c]: The docs just say that `sni_callback` has to return None. What's this extra context setting stuff about? Should it be removed?

**Review by njsmith — COMMENTED** [review]
I pushed some prose changes, and here are some comments on the code. Overall looks good, but there are a few small cleanups to consider, and a few points where I have questions.

File: Lib/ssl.py (hunk 3 above)
Comment (tiran) [reply Lib/ssl.py]: Ah right, ``callable`` was re-added in Python 3.2. I'll change it.

File: Lib/ssl.py (hunk 4 above)
Comment (tiran) [reply Lib/ssl.py]: I don't mind to use the method here, too.

File: Lib/ssl.py (hunk 5 above)
Comment (tiran) [reply Lib/ssl.py:358]: I'd prefer to keep this a method. It makes it easier to modify the behavior in subclasses or in tests.

File: Modules/_ssl.c (hunk 6 above)
Comment (tiran) [reply Modules/_ssl.c:4245]: It's not new: https://docs.python.org/3/library/ssl.html#ssl.SSLContext.protocol

I just moved it from Python code into C code so I can access the value from C methods. The new setter for the callback uses the value to prevent setting a SNI callback for a client-only context.

File: Modules/_ssl.c (hunk 7 above)
Comment (tiran) [reply Modules/_ssl.c]: I added the code for future compatibility with PEP 543. Rather than modifying socket.context, the callback can return a new context and let the API take care of the rest.

By the way it doesn't have to be None. Other values signal an error:

> The server_name_callback function must return None to allow the TLS negotiation to continue. If a TLS failure is required, a constant ALERT_DESCRIPTION_* can be returned. Other return values will result in a TLS fatal error with ALERT_DESCRIPTION_INTERNAL_ERROR.

File: Modules/_ssl.c (hunk 6 above)
Comment (njsmith) [reply Modules/_ssl.c:4245]: Ah, gotcha.

File: Modules/_ssl.c (hunk 7 above)
Comment (njsmith) [reply Modules/_ssl.c]: Ok yeah, I think adding new undocumented features in b2 for partial compatibility with an unfinished PEP is probably a bad plan, so thanks for removing that :-). It'll be easy to add back later.

**Review by njsmith — APPROVED** [review]

**Comment by tiran** [comment]
The context special handling is pining for the fjords.

**Comment by mattip** [comment]
Why are there no tests for the new function name, the tests in the stdlib all refer to `set_servername_callback` ?
